File: rsa.js
BigInt.prototype.log2 = function () {
    return (this  > 1n) ? (1n + (this/2n).log2()) : 0n
}

String.prototype.encode = function (rsa) {    
    let n = BigInt(rsa.n), e = BigInt(rsa.e), result = ""
    let M = this.toBigInt()
    let mBinary = M.toString(2)

    

    let bits = (n-1n).log2()

    let l_mBinary = BigInt(mBinary.length)
    
    if (l_mBinary <= bits) {
        let C = powMod(M, e, n)        
        console.log(C);
        
        result = C.toWord()        
    } else {
        let count = (l_mBinary % bits == 0n) ? (l_mBinary / bits) : (l_mBinary / bits + 1n)

        let segment = []

        if (l_mBinary % bits == 0n) {
            for (let i = 0n; i < count; i++) {
                segment[i] = mBinary.substring(Number(i*bits), Number((i+1n)*bits))
            }
        } else {
            segment[0] = mBinary.substring(0, Number(l_mBinary % bits))
            for (let i = 0n; i < count; i++) {
                segment[i] = mBinary.substring(Number(l_mBinary % bits + (i - 1n) * bits), Number(l_mBinary % bits + i * bits))
            }
        }
        

        let listDecimal = [], listBinary = [],C = [] 
        listDecimal.length = Number(count)
        listBinary.length = Number(count)
        C.length = Number(count)
                

        for (let i = 0n; i < count; i++) {
            listDecimal[i] = BigInt("0b"+segment[i])            
            C[i] = powMod(listDecimal[i], e, n)
            listBinary[i] = C[i].toString(2)
            
            let list_l = BigInt(listBinary[i].length)
            if (list_l < bits + 1n) {
                let repeat = []
                repeat.length = Number(bits + 1n - list_l)
                repeat.fill(0)
                listBinary[i] = repeat.join("") + listBinary[i]
            }
            result += listBinary[i]
        }
        
        let resultDecimal = BigInt("0b"+result)
        
        result = resultDecimal.toWord()
    }
    
    return result
}

String.prototype.decode = function (rsa) {
    let n = BigInt(rsa.n), d = BigInt(rsa.d)    
    
    let result = "", C = this.toBigInt()
    let cBinary = C.toString(2)

    let bits = (n-1n).log2()


    let l_cBinary = BigInt(cBinary.length)
    
    
    if (l_cBinary <= bits) {
        let M = powMod(C, d, n)
        
        result = M.toWord()
    } else {
        let count = (l_cBinary % (bits + 1n) == 0n) ? (l_cBinary / (bits + 1n)) : (l_cBinary / (bits + 1n) + 1n)
        
        let segment = []
        

        if (l_cBinary % (bits + 1n) == 0n) {
            for (let i = 0n; i < count; i++) {
                segment[i] = cBinary.substring(Number(i*(bits + 1n)), Number((i+1n)*(bits + 1n)))
            }
        } else {
            segment[0] = cBinary.substring(0, Number(l_cBinary % (bits + 1n)))
            for (let i = 0n; i < count; i++) {
                segment[i] = cBinary.substring(Number(l_cBinary % (bits + 1n) + (i - 1n) * (bits + 1n)), Number(l_cBinary % (bits + 1n) + i * (bits + 1n)))
            }
        }

        let listDecimal = [], listBinary = [],M = [] 
        listDecimal.length = Number(count)
        listBinary.length = Number(count)
        M.length = Number(count)

        for (let i = 0n; i < count; i++) {
            listDecimal[i] = BigInt("0b"+segment[i])            
            M[i] = powMod(listDecimal[i], d, n)
            listBinary[i] = M[i].toString(2)

            let list_l = BigInt(listBinary[i].length)

            if (list_l < bits) {
                let repeat = []
                repeat.length = Number(bits - list_l)
                repeat.fill(0)
                listBinary[i] = repeat.join("") + listBinary[i]
            }
            
            result += listBinary[i]
        }

        let resultDecimal = BigInt("0b"+result)
        result = resultDecimal.toWord()
    
    }

    return result
}

String.prototype.toBigInt = function (n) {
    if (n==null) n = 94n
    n = BigInt(n)

    let l = BigInt(this.length - 1)
    let result = 0n
    for (let i = 0n; i < BigInt(this.length); i++) {
        const c = this.charAt(Number(i))
    
        let temp = n
        temp = temp ** l

        if (c == "~") 
            temp = 0n
        else 
            temp = temp * (BigInt(c.charCodeAt(0)) - 32n)
        result = result + temp
        l--        
    }

    return result
}

BigInt.prototype.toWord = function (n) {
    let number = BigInt(this)

    if (n==null) n = 94n 

    let word = ""
    let temp 

    while (number > 0) {
        temp = number%n + 32n
        if (temp == 32n) 
            temp = 126n;
        word += String.fromCharCode(Number(temp))
        number = number/n
    }

    return word.split("").reverse().join("")
}

function powMod(base, exp, mod) {
    if (exp == 0n) return 1n;
    if (exp % 2n == 0) {
        return powMod(base, (exp / 2n), mod) ** 2n % mod;
    }
    else {
        return (base * powMod(base, (exp - 1n), mod)) % mod;
    }
}

function encode(raw, rsa_public) {
    let rawSplit = raw.split(" ")
    let cipherSplit = []
    rawSplit.forEach(word => {        
        word = encodeURI(word)        
        cipherSplit.push(word.encode(rsa_public))
        
    });
    return cipherSplit.join(" ")
}

function decode(cipher, rsa_private) {
    let cipherSplit = cipher.split(" ")
    let rawSplit = []
    cipherSplit.forEach(word => {
        rawSplit.push(decodeURI(word.decode(rsa_private)))
    });
    return rawSplit.join(" ")
}

function privateKey(rsa_public) {
    // Tìm p & q
    let n = rsa_public.n
    let e = rsa_public.e
    let sqrtN = Math.floor(Math.sqrt(n))
    let p = sqrtN, q

    while (p != 1) {
        if (n % p == 0) break
        p = p - 1
    }

    q = Math.floor(n / p)

    // m = (p-1) * (q-1);
    let m = (p - 1) * (q - 1)

    let xa = 1, ya = 0, xb = 0, yb = 1, temp = m

    while (m != 0) {
        let z = Math.floor(e / m)
        let r = e % m
        e = m;
        m = r;
        let xr = xa - z * xb
        let yr = ya - z * yb
        xa = xb;
        ya = yb;
        xb = xr;
        yb = yr;
    }

    if (xa < 0) {
        xa = xa + temp
    }

    return {p, q, d: xa, m: temp }

}

function cryptanalysis(cipher, rsa) {
    rsa = {
        ...rsa,
        ...privateKey(rsa)
    }
    return { raw: decode(cipher, rsa), rsa }
}

const rsa_public = {
    e: 17,
    n: 3233
}

const rsa_private = {
    n: 3233,
    d: 2753
}

console.log(cryptanalysis(encode("Nguyễn, QuangNgọc Thế Vinh", rsa_public), rsa_public));

File: RSA.java
import java.math.*;
import java.util.*;

public class RSA {

	static Scanner sc = new Scanner(System.in);

	public static void main(String[] args) {
		// Bộ thử RSA
		// p = 51343
		// q = 62981
		// n = p * q = 3233633483
		// m = (p-1) * (q-1) = 3233519160
		// Public key e = 65537
		// Private key d = 2854695593

		int choose = 0;
		do {
			System.out.println("****************************************");
			System.out.println("RSA MENU");
			System.out.println("1. Encryption");
			System.out.println("2. Cryptanalysis");
			System.out.println("0. Exit");
			System.out.println("****************************************");
			System.out.print("Choose function: ");
			choose = sc.nextInt();
			switch (choose) {
			case 1:
				encryption();
				break;
			case 2:
				cryptanalysis();
				break;
			}
		} while (choose != 0);
		System.out.println("Bye. See you again.");
		sc.close();
	}

	// Chức năng 1 mã hóa encryption
	static void encryption() {
		System.out.println("\n****************************************");
		System.out.println("Function 1: Encryption\n");

		System.out.print("Input n = ");
		sc.nextLine();
		String nString = sc.nextLine();
		BigInteger n = new BigInteger(nString);

		System.out.print("Input public key e = ");
		String eString = sc.nextLine();
		BigInteger e = new BigInteger(eString);

		// Nhập bản rõ để mã hóa
		System.out.println("Input plain text:");
		String plaintext = sc.nextLine();

		// Chuyển đổi bản rõ sang bản mã
		String ciphertext = new String();
		String[] list = plaintext.split(" ");
		for (int i = 0; i < list.length; i++) {
			ciphertext += encode(list[i], e, n) + " ";
		}
		ciphertext = ciphertext.trim();
		System.out.println("Cipher text:\n" + ciphertext);
		System.out.println("****************************************\n");
	}

	// Chức năng 2 giải khóa cryptanalysis
	static void cryptanalysis() {
		System.out.println("\n****************************************");
		System.out.println("Function 2: Cryptanalysis\n");

		System.out.print("Input n = ");
		sc.nextLine();
		String nString = sc.nextLine();
		BigInteger n = new BigInteger(nString);

		System.out.print("Input public key e = ");
		String eString = sc.nextLine();
		BigInteger e = new BigInteger(eString);

		// Phân tích bộ mã
		BigInteger d = privateKey(e, n);

		// Nhập bản mã để giải mã
		System.out.println("\nInput cipher text:");
		String ciphertext = sc.nextLine();

		// Chuyển đổi bản mã sang bản rõ
		String plaintext = new String();
		String[] list = ciphertext.split(" ");
		for (int i = 0; i < list.length; i++) {
			plaintext += decode(list[i], d, n) + " ";
		}
		plaintext = plaintext.trim();
		System.out.println("Plain text:\n" + plaintext);
		System.out.println("****************************************\n");
	}

	// Mã hóa một word
	static String encode(String word, BigInteger e, BigInteger n) {
		String result = new String();
		BigInteger M = wordToNumber(word); // Đổi word bản rõ sang số hệ 10
		String mBinary = decimalToBinary(M); // Đổi số hệ 10 sang binary

		// Tính số bits mỗi đoạn có thể mã hóa bits = log2(n)
		int bits = (int) (log(n.subtract(BigInteger.ONE)) / log(new BigInteger("2")));

		if (mBinary.length() <= bits) {
			// Độ dài word bản rõ đủ để mã hóa bởi n
			BigInteger C = powMod(M, e, n); // Tính C = M^e mod n
			result = numberToWord(C); // Đổi số hệ 10 sang word bản mã
		} else {
			// Độ dài word bản rõ dài quá n, phải cắt thành các đoạn
			// Tính số phân đoạn count
			int count = (mBinary.length() % bits == 0) ? mBinary.length() / bits : mBinary.length() / bits + 1;

			// Cắt từng phân đoạn
			String[] segment = new String[count];
			if (mBinary.length() % bits == 0) {
				for (int i = 0; i < count; i++) {
					segment[i] = mBinary.substring(i * bits, (i + 1) * bits);
				}
			} else {
				segment[0] = mBinary.substring(0, mBinary.length() % bits);
				for (int i = 1; i < count; i++) {
					segment[i] = mBinary.substring(mBinary.length() % bits + (i - 1) * bits,
							mBinary.length() % bits + i * bits);
				}
			}

			// Mã hóa từng phân đoạn
			BigInteger[] listDecimal = new BigInteger[count];
			String[] listBinary = new String[count];
			BigInteger[] C = new BigInteger[count];
			for (int i = 0; i < count; i++) {
				listDecimal[i] = binarytoDecimal(segment[i]); // Chuyển các phân đoạn từ binary sang số hệ 10
				C[i] = powMod(listDecimal[i], e, n); // Tính C[i] = listDecimal[i]^e mod n
				listBinary[i] = decimalToBinary(C[i]); // Chuyển các phân đoạn đã mã hóa từ số hệ 10 sang binary
				if (listBinary[i].length() < bits + 1) {
					// Nếu số bits các phân đoạn không bằng bits + 1
					char[] repeat = new char[bits + 1 - listBinary[i].length()];
					Arrays.fill(repeat, '0');
					// Thêm các số 0 đằng trước cho đủ số bits
					listBinary[i] = new String(repeat) + listBinary[i];
				}
				result += listBinary[i]; // Gộp các phân đoạn binary lại
			}

			BigInteger resultDecimal = binarytoDecimal(result); // Đổi result từ binary sang số hệ 10
			result = numberToWord(resultDecimal); // Đổi số hệ 10 sang word bản mã
		}
		return result;
	}

	// Giải mã một word
	static String decode(String word, BigInteger d, BigInteger n) {
		String result = new String();
		BigInteger C = wordToNumber(word); // Đổi word bản mã sang số hệ 10
		String cBinary = decimalToBinary(C); // Đổi số hệ 10 sang binary

		// Tính số bits mỗi đoạn có thể biểu diễn
		// bits = log2(n)
		int bits = (int) (log(n.subtract(BigInteger.ONE)) / log(new BigInteger("2")));

		if (cBinary.length() <= bits) {
			// Độ dài word bản mã đủ để giải mã bởi n
			BigInteger M = powMod(C, d, n); // Tính M = C^d mod n
			result = numberToWord(M); // Đổi số hệ 10 sang word bản rõ
		} else {
			// Độ dài word bản mã dài quá n, phải cắt thành các đoạn
			// Tính số phân đoạn count
			int count = (cBinary.length() % (bits + 1) == 0) ? cBinary.length() / (bits + 1)
					: cBinary.length() / (bits + 1) + 1;

			// Cắt từng phân đoạn
			String[] segment = new String[count];
			if (cBinary.length() % (bits + 1) == 0) {
				for (int i = 0; i < count; i++) {
					segment[i] = cBinary.substring(i * (bits + 1), (i + 1) * (bits + 1));
				}
			} else {
				segment[0] = cBinary.substring(0, cBinary.length() % (bits + 1));
				for (int i = 1; i < count; i++) {
					segment[i] = cBinary.substring(cBinary.length() % (bits + 1) + (i - 1) * (bits + 1),
							cBinary.length() % (bits + 1) + i * (bits + 1));
				}
			}

			// Giải mã từng phân đoạn
			BigInteger[] listDecimal = new BigInteger[count];
			String[] listBinary = new String[count];
			BigInteger[] M = new BigInteger[count];
			for (int i = 0; i < count; i++) {
				listDecimal[i] = binarytoDecimal(segment[i]); // Chuyển các phân đoạn từ binary sang số hệ 10
				M[i] = powMod(listDecimal[i], d, n); // Tính M[i] = listDecimal[i]^d mod n
				listBinary[i] = decimalToBinary(M[i]); // Chuyển các phân đoạn đã giải mã từ số hệ 10 sang binary
				if (listBinary[i].length() < bits) {
					// Nếu số bits các phân đoạn không bằng bits
					char[] repeat = new char[bits - listBinary[i].length()];
					Arrays.fill(repeat, '0');
					// Thêm các số 0 đằng trước cho đủ số bits
					listBinary[i] = new String(repeat) + listBinary[i];
				}
				result += listBinary[i]; // Gộp các phân đoạn binary lại
			}

			BigInteger resultDecimal = binarytoDecimal(result); // Đổi result từ binary sang số hệ 10
			result = numberToWord(resultDecimal); // Đổi số hệ 10 sang word bản rõ
		}
		return result;
	}

	// Phá khóa - Tìm private key d
	static BigInteger privateKey(BigInteger e, BigInteger n) {
		System.out.println("\nAnalysis results:");
		// Phân tích n = p * q
		BigInteger sqrtN = sqrt(n);
		BigInteger p = sqrtN, q;
		while (p.equals(BigInteger.ONE) == false) {
			if (n.mod(p).compareTo(BigInteger.ZERO) == 0)
				break;
			p = p.subtract(BigInteger.ONE);
		}
		q = n.divide(p);
		System.out.println("p = " + p);
		System.out.println("q = " + q);

		// m = (p-1) * (q-1);
		p = p.subtract(BigInteger.ONE);
		q = q.subtract(BigInteger.ONE);
		BigInteger m = p.multiply(q);
		System.out.println("m = (p-1) * (q-1) = " + m);
		BigInteger temp = m;

		// Thuật toán Euclid
		// Giải x * a + y * b = 1
		// d * e + y * m = 1
		BigInteger xa = BigInteger.ONE;
		BigInteger ya = BigInteger.ZERO;
		BigInteger xb = BigInteger.ZERO;
		BigInteger yb = BigInteger.ONE;

		while (m.equals(BigInteger.ZERO) == false) {
			BigInteger z = e.divide(m);
			BigInteger r = e.mod(m);
			e = m;
			m = r;
			BigInteger xr = xa.subtract(z.multiply(xb));
			BigInteger yr = ya.subtract(z.multiply(yb));
			xa = xb;
			ya = yb;
			xb = xr;
			yb = yr;
		}
		if (xa.compareTo(BigInteger.ZERO) < 0) {
			xa = xa.add(temp);
		}

		System.out.println("Private key d = " + xa);
		return xa;
	}

	// Tính y = x^a mod n
	static BigInteger powMod(BigInteger x, BigInteger a, BigInteger n) {
		String binaryA = decimalToBinary(a); // Đổi a sang binary
		BigInteger y = BigInteger.ONE;
		for (int i = 0; i < binaryA.length(); i++) {
			y = y.multiply(y);
			y = y.mod(n);
			if (binaryA.charAt(i) == '1') {
				y = y.multiply(x);
				y = y.mod(n);
			}
		}
		return y;
	}

	// Đổi word (hệ 27) sang số hệ 10
	static BigInteger wordToNumber(String word) {
		int length = word.length() - 1;
		BigInteger result = BigInteger.ZERO;
		for (int i = 0; i < word.length(); i++) {
			BigInteger temp = new BigInteger("27");
			temp = temp.pow(length); // 27 ^ length
			temp = temp.multiply(new BigInteger(Integer.toString((int) word.charAt(i) - 96)));
			result = result.add(temp);
			length--;
		}
		return result;
	}

	// Đổi số hệ 10 sang word (hệ 27)
	static String numberToWord(BigInteger number) {
		String word = new String();
		long temp;
		while (number.compareTo(BigInteger.ZERO) > 0) {
			temp = number.mod(new BigInteger("27")).intValue() + 96;
			word += (char) temp;
			number = number.divide(new BigInteger("27"));
		}
		StringBuilder result = new StringBuilder(word);
		return new String(result.reverse());
	}

	// Đổi số nhị phân sang số hệ 10
	static BigInteger binarytoDecimal(String binary) {
		BigInteger result = BigInteger.ZERO;
		BigInteger base = BigInteger.ONE;
		for (int i = binary.length() - 1; i >= 0; i--) {
			if (binary.charAt(i) == '1')
				result = result.add(base);
			base = base.multiply(new BigInteger("2"));
		}
		return result;
	}

	// Đổi số hệ 10 sang số nhị phân
	static String decimalToBinary(BigInteger decimal) {
		String word = new String();
		while (decimal.compareTo(BigInteger.ZERO) > 0) {
			word += String.valueOf(decimal.mod(new BigInteger("2")));
			decimal = decimal.divide(new BigInteger("2"));
		}
		StringBuilder result = new StringBuilder(word);
		return new String(result.reverse());
	}

	// Tính căn bậc hai của một số BigInteger
	static BigInteger sqrt(BigInteger n) {
		BigInteger a = BigInteger.ONE;
		BigInteger b = n.shiftRight(5).add(BigInteger.valueOf(8));
		while (b.compareTo(a) >= 0) {
			BigInteger mid = a.add(b).shiftRight(1);
			if (mid.multiply(mid).compareTo(n) > 0) {
				b = mid.subtract(BigInteger.ONE);
			} else {
				a = mid.add(BigInteger.ONE);
			}
		}
		return a.subtract(BigInteger.ONE);
	}

	// Tính logarit của một số BigInteger
	static double log(BigInteger val) {
		if (val.signum() < 1)
			return val.signum() < 0 ? Double.NaN : Double.NEGATIVE_INFINITY;
		int blex = val.bitLength() - 977;
		if (blex > 0)
			val = val.shiftRight(blex);
		double res = Math.log(val.doubleValue());
		return blex > 0 ? res + blex * Math.log(2) : res;
	}

}